### ISSUE TITLE
Encoding error when adding description lines in Propal or Invoice

### ISSUE THREAD

**Issue by vijdz** [Dolibarr/dolibarr#23750]
### Bug

When creating a new propal or invoice, in French language  (using special charachters like "é" or "è"), the result appears with encoding errors. I can't see any useful output in the main log file.  
Confirmed with firefox or chrome, on Linux or Windows.

### Environment Version

Current branch 16.0

### Environment OS

Ubuntu 20.04

### Environment Web server

Nginx 1.18

### Environment PHP

PHP 7.4

### Environment Database

PostgreSQL 12.13

### Environment URL(s)

/comm/propal, /compta/facture

### Expected and actual behavior

Special characters worked fine until upgrade from 15 to 16

### Steps to reproduce the behavior

For example, adding a new "product" line to a propal with text "ééé" will result in " Ã©Ã©Ã©".

**Comment by Daviid-P** [comment]
Check conf.php

```php
$dolibarr_main_db_character_set = 'utf8';
$dolibarr_main_db_collation = 'utf8_general_ci';
``` 

and run these in your database.

```SQL
SHOW TABLE STATUS;

SHOW VARIABLES LIKE  'char%';
```

**Comment by vijdz** [comment]
I had these in the conf.php
```
$dolibarr_main_db_character_set='UTF8';
$dolibarr_main_db_collation='en_US.UTF-8';
```
It does not change anything to set this to utf8_general_ci.  
Regarding psql, I don't understand if I'm supposed to replace something in your commands, but these don't look like valid SQL (in postgres at least) : 
``` 
dolibarr=# SHOW TABLE STATUS;
ERROR:  syntax error at or near "TABLE"
LINE 1: SHOW TABLE STATUS;
             ^
dolibarr=# 
dolibarr=# SHOW VARIABLES LIKE  'char%';
ERROR:  syntax error at or near "LIKE"
LINE 1: SHOW VARIABLES LIKE  'char%';
                       ^
```

But I have this :  
```
dolibarr=# SHOW SERVER_ENCODING;
 server_encoding 
-----------------
 UTF8
(1 row)
```

**Comment by vijdz** [comment]
I checked the "llx_propal" table and everything seems to be in UTF8 as it should... 
To me this is a front-end bug, or something went wrong in the back-end with the POST encoding conversion ?

**Comment by vijdz** [comment]
Response header when adding a line with special char to a an invoice :  
```
HTTP/1.1 200 OK
Content-Length: 120385
Cache-Control: no-store, no-cache, must-revalidate
Connection: keep-alive
Content-Type: text/html; charset=UTF-8
Date: Tue, 07 Feb 2023 08:34:24 GMT
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Pragma: no-cache
Server: nginx/1.18.0 (Ubuntu)
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
```

I believe evertyhing is OK with the server, nothing has changed (except ubuntu security updates) since upgrade from 15.0

**Comment by Daviid-P** [comment]
@vidlb 
Whats the `<meta charset="utf-8">` element in your html response?

And maybe `<html lang="es">`

**Comment by vijdz** [comment]
In the html response :  
```html
<!doctype html>
<html lang="fr">
<head>
<meta charset="utf-8">
...

```

**Comment by Daviid-P** [comment]
Then I have no idea what's wrong.
My last idea is to modify the files to log the values to file before they're printed to the browser. Or try to log the hex values instead and see if everything is as it should be.

**Comment by AndreasSt-2022** [comment]
Does it help to set MAIN_RESTRICTHTML_ONLY_VALID_HTML to "0" in the extended configuration (setup other)?

**Comment by vijdz** [comment]
I just tried and it prevents me from adding a descr line.  
Chars `à & é` are replaced by `&agrave; &amp; &eacute;` in the field, but it does not add the line.

**Comment by vijdz** [comment]
Here is the debug log when I try to insert a line to an invoice description with special char :  
```
2023-02-22 15:17:24 DEBUG   193.48.189.250    sql=INSERT INTO llx_facturedet (fk_facture, fk_parent_line, label, description, qty, vat_src_code, tva_tx, localtax1_tx, localtax2_tx, localtax1_type, localtax2_type, fk_product, product_type, remise_percent, subprice, ref_ext, fk_remise_except, date_start, date_end, fk_code_ventilation,  rang, special_code, fk_product_fournisseur_price, buy_price_ht, info_bits, total_ht, total_tva, total_ttc, total_localtax1, total_localtax2, situation_percent, fk_prev_id, fk_unit, fk_user_author, fk_user_modif, fk_multicurrency, multicurrency_code, multicurrency_subprice, multicurrency_total_ht, multicurrency_total_tva, multicurrency_total_ttc) VALUES (155, null, null, '&Atilde;&copy;&Atilde;&copy;&Atilde;&nbsp;&Atilde;&nbsp;', 1, '', 20, 0, 0, '0', '0', null, 0, 0, 13, '', null, null, null, 0, 1, 0, null, 0, '0', 13, 2.6, 15.6, 0, 0, 100, null, NULL, 1, 1, 0, 'EUR', 13, 13, 2.6, 15.6)
```
The text inserted in the database is `'&Atilde;&copy;&Atilde;&copy;&Atilde;&nbsp;&Atilde;&nbsp;'`.  
Is it expected ?  
Anyone in the dev team ? What should I do to help you locate the source of this problem ?

**Comment by vijdz** [comment]
The problem seems to apply to any type of comment / description linked to a product or company.  
But it does not occur if I try to add a company with special characters for example, and the accents are preserved - I can see it in the debug log.

**Comment by jtraulle** [comment]
I have the exact same problem.

```shell
MariaDB [(none)]> SHOW VARIABLES LIKE  'char%';
+--------------------------+----------------------------+
| Variable_name            | Value                      |
+--------------------------+----------------------------+
| character_set_client     | utf8                       |
| character_set_connection | utf8                       |
| character_set_database   | utf8mb4                    |
| character_set_filesystem | binary                     |
| character_set_results    | utf8                       |
| character_set_server     | utf8mb4                    |
| character_set_system     | utf8                       |
| character_sets_dir       | /usr/share/mysql/charsets/ |
+--------------------------+----------------------------+
```

Dolibarr database, tables and text columns are in `utf8` / `utf8_unicode_ci` collation.

```php
$dolibarr_main_db_type='mysqli';
$dolibarr_main_db_character_set='utf8';
$dolibarr_main_db_collation='utf8_unicode_ci';
```

Problem have appeared when upgrading to 16.

**Comment by jtraulle** [comment]
> Does it help to set MAIN\_RESTRICTHTML\_ONLY\_VALID\_HTML to "0" in the extended configuration (setup other)?

Thank you @AndreasSt-2022, you put me in the way.

The constant `MAIN_RESTRICTHTML_ONLY_VALID_HTML` were defined to 1 in my installation and that was what caused the accented characters to be replaced wrongly.

**Comment by vijdz** [comment]
Well I just tried and it worked.
Also, to make it also work for templates exported as .odt files, I had to set MAIN_RESTRICTHTML_REMOVE_ALSO_BAD_ATTRIBUTES to 0.

I don't understand why I had a different error the first time I tried. May be something was patched in 16.04 ? Or I was doing something wrong.

The thing is this variable / setting is one of the suggested configuration for a more secure installation. I think it was advertised like this in the security section of the admin module (well now I remember it was marked as experimental).

May be the docs should state that this is incompatible with some languages for now.
Closing it now !
